### PR TITLE
List(...) optimization to avoid intermediate array

### PR DESCRIPTION
**Pull request by KuceraMartin**

### CODE REVIEW

**Comment by KuceraMartin** [comment]
todo:
- [ ] fix lists of primitive types
- [ ] limit for nested lists
- [ ] more tests

**Comment by mbovel** [comment]
Hi @KuceraMartin! Would you like to continue working on this PR during the next issue spree? Do you think the TODOs left are easy and could be interesting for beginners to follow along?

**Comment by KuceraMartin** [comment]
> Hi @KuceraMartin! Would you like to continue working on this PR during the next issue spree? Do you think the TODOs left are easy and could be interesting for beginners to follow along?

Hi @mbovel, actually I'm not sure if this is suitable for another spree.

1. The first todo might be interesting but maybe too easy?
2. For the second one I'm not sure—if I remember correctly, the problem with the Scala 2 implementation was that it was calling `super.transform` recursively to handle the nesting, which we cannot do with the Scala 3 mini phases? But then I think @smarter mentioned he wasn't sure if we should re-implement this nested limit at all.
3. Writing more tests is probably not very interesting.

I was planning to finish the PR by myself. If it doesn't get picked for the spree, I should have time to do it next week.

File: compiler/src/dotty/tools/dotc/transform/ArrayApply.scala [hunk 1]
```diff
@@ -22,27 +19,71 @@ class ArrayApply extends MiniPhase {
...
+  private def isSeqApply(tree: Tree)(using Context): Boolean =
+    isListApply(tree) || tree.symbol == defn.SeqModule_apply && appliedCore(tree).match
+      case Select(qual, _) =>
+        val sym = qual.symbol
+        sym == defn.SeqModule
+        || sym == defn.SeqModuleAlias
+        || sym == defn.CollectionSeqType.symbol.companionModule
+      case _ => false
+
+  private def seqApplyArgsOrNull(tree: Apply)(using Context): JavaSeqLiteral | Null =
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/transform/ArrayApply.scala:78]: We usually avoid nulls like that one. Maybe define a `SeqApply(args)` extractor that covers both `isSeqApply` and `seqApplyArgsOrNull`.

Example use:
```diff
   override def prepareForApply(tree: Apply)(using Context): Context =
-    if isSeqApply(tree) then
-      val args = seqApplyArgsOrNull(tree)
-      if args != null then
-        ctx.fresh.setProperty(TransformListApplyBudgetKey, transformListApplyBudget - args.elems.length)
-     else ctx
-    else ctx
+    tree match
+      case SeqApply(args) => 
+        ctx.fresh.setProperty(TransformListApplyBudgetKey, transformListApplyBudget - args.elems.length)
+      case _ => 
+        ctx
```

File: compiler/test/dotty/tools/backend/jvm/ArrayApplyOptTest.scala [hunk 2]
```diff
@@ -160,4 +161,79 @@ class ArrayApplyOptTest extends DottyBytecodeTest {
...
+  @Test def testSeqApplyAvoidsIntermediateArray3 = {
+    checkApplyAvoidsIntermediateArray("scala.collection.Seq"):
+      """import scala.collection.immutable.{ ::, Nil }, scala.collection.Seq
+        |class Foo {
+        |  def meth1: Seq[String] = Seq("1", "2", "3")
+        |  def meth2: Seq[String] = new ::("1", new ::("2", new ::("3", Nil)))
+        |}
+    """.stripMargin
+  }
+
```
Comment (nicolasstucki) [inline compiler/test/dotty/tools/backend/jvm/ArrayApplyOptTest.scala:203]: Also add tests for 

* `List[Object]("1", "2", "3", "4", "5", "6", List[Object]()) // 7 cons w/ 6 string heads + 1 `
* ` List[Object]("1", "2", "3", "4", "5", List[Object]("6", "7")) // wrap 2`

**Review by nicolasstucki — CHANGES_REQUESTED** [review]

**Review by nicolasstucki — APPROVED** [review]

**Pull request merged**